File: public/examples/wave-equation/js/wave.js
/* Deriving the wave equation -- a chain of beads, evolved exactly.
 *
 * The model IS the derivation: N beads of mass m = mu*dx joined by string
 * under tension T. Newton on bead i gives
 *
 *     y_i'' = c^2 * (y_{i+1} - 2 y_i + y_{i-1}) / dx^2,   c^2 = T/mu
 *
 * and the bracket is the second difference -- sampled curvature. Send
 * dx -> 0 and it becomes y_tt = c^2 y_xx. The page lets you watch that
 * limit happen by raising N.
 *
 * Evolution is EXACT, not stepped: the chain's eigenmodes are known
 * (y_i ~ sin(k pi i/(N+1)), w_k = (2c/dx) sin(k pi / (2(N+1)))), so a
 * pluck is decomposed once by discrete sine transform and every later
 * shape is a pure function of t. No integrator, no drift, scrub-safe.
 * Verified against an independent Verlet integration to 2.4e-9 before
 * this page was built.
 *
 * One honest wrinkle the check surfaced: the chain's SHORTEST waves never
 * reach c -- their speed pins at 2/pi of c no matter how large N gets.
 * What the limit fixes is every wave a smooth pluck actually contains,
 * which is why the fidelity readout is energy-weighted over the pluck's
 * own spectrum (and why it approaches 100% as N grows).
 */
"use strict";

// ---- physics + presentation config (data) ------------------------------
const CONFIG = {
  L: 10,                      // string length, world units
  amp: 1.3,                   // pluck amplitude
  focusFrac: 0.45,            // which bead the derivation panel dissects
  world: { x0: -0.5, x1: 10.5, y0: -2.2, y1: 2.2 },  // 11 x 4.4 == 1000 x 400
  viewBox: { w: 1000, h: 400 },
};

const { sx, sy, wx } = Lab.world(CONFIG.world, CONFIG.viewBox);
const el = Lab.el;
const W = CONFIG.world, VB = CONFIG.viewBox;
const L = CONFIG.L;

const state = { N: 24, c: 1.5, pos: 0.3, t: 0, playing: true, rate: 1, step: 1 };
const scene = document.getElementById("scene");
const svg = document.getElementById("svg");

// ---- the model: modes of the discrete chain ----------------------------
let model = null;

function rebuild() {
  const N = state.N, c = state.c;
  const dx = L / (N + 1);
  // sine table S[k][i], k,i = 1..N
  const S = [];
  for (let k = 1; k <= N; k++) {
    const row = [];
    for (let i = 1; i <= N; i++) row.push(Math.sin(k * Math.PI * i / (N + 1)));
    S.push(row);
  }
  // mode frequencies
  const w = [];
  for (let k = 1; k <= N; k++) w.push((2 * c / dx) * Math.sin(k * Math.PI / (2 * (N + 1))));
  // triangle pluck at pos, zero velocity
  const p = Math.min(0.92, Math.max(0.08, state.pos));
  const y0 = [];
  for (let i = 1; i <= N; i++) {
    const x = i / (N + 1);
    y0.push(CONFIG.amp * (x <= p ? x / p : (1 - x) / (1 - p)));
  }
  // discrete sine transform -> mode amplitudes
  const a = [];
  for (let k = 1; k <= N; k++) {
    let sum = 0;
    for (let i = 1; i <= N; i++) sum += y0[i - 1] * S[k - 1][i - 1];
    a.push((2 / (N + 1)) * sum);
  }
  // energy-weighted mean phase-speed ratio of THIS pluck's spectrum
  let num = 0, den = 0;
  for (let k = 1; k <= N; k++) {
    const th = k * Math.PI / (2 * (N + 1));
    const E = (a[k - 1] * w[k - 1]) ** 2;
    num += E * Math.sin(th) / th;
    den += E;
  }
  model = { N, dx, S, w, a, fidelity: den > 0 ? num / den : 1,
            focus: Math.max(2, Math.min(N - 1, Math.round(CONFIG.focusFrac * (N + 1)))) };
}

/** Bead displacements at time t -- a pure function, nothing accumulates. */
function shapeAt(t) {
  const { N, S, w, a } = model;
  const y = new Array(N).fill(0);
  for (let k = 1; k <= N; k++) {
    const ck = a[k - 1] * Math.cos(w[k - 1] * t);
    if (Math.abs(ck) < 1e-12) continue;
    const row = S[k - 1];
    for (let i = 0; i < N; i++) y[i] += ck * row[i];
  }
  return y;
}

// ---- render ------------------------------------------------------------
function render() {
  const { N, dx, focus } = model;
  const y = shapeAt(state.t);
  scene.textContent = "";

  // rest line + walls
  scene.appendChild(el("line", { x1: sx(0), y1: sy(0), x2: sx(L), y2: sy(0),
    stroke: "var(--wv-rest)", "stroke-width": 1, "stroke-dasharray": "3 6", opacity: 0.6 }));
  for (const xw of [0, L]) {
    scene.appendChild(el("line", { x1: sx(xw), y1: sy(-1.9), x2: sx(xw), y2: sy(1.9),
      stroke: "var(--wv-wall)", "stroke-width": 3 }));
  }

  // the string through the beads (fixed ends included)
  let pts = sx(0) + "," + sy(0);
  for (let i = 1; i <= N; i++) pts += " " + sx(i * dx) + "," + sy(y[i - 1]);
  pts += " " + sx(L) + "," + sy(0);
  scene.appendChild(el("polyline", { points: pts, fill: "none",
    stroke: "var(--wv-string)", "stroke-width": 2, "stroke-linejoin": "round" }));

  // beads
  const r = Math.max(2.4, 7 - N * 0.05);
  for (let i = 1; i <= N; i++) {
    scene.appendChild(el("circle", { cx: sx(i * dx), cy: sy(y[i - 1]), r,
      fill: i === focus && state.step >= 1 ? "var(--wv-focus)" : "var(--wv-bead)" }));
  }

  // derivation overlays on the focus bead
  if (state.step >= 2) {
    const j = focus;
    const [xl, xj, xr] = [(j - 1) * dx, j * dx, (j + 1) * dx];
    const [yl, yj, yr] = [y[j - 2], y[j - 1], y[j]];
    // its two string segments, emphasized: these carry the tension
    scene.appendChild(el("polyline", {
      points: `${sx(xl)},${sy(yl)} ${sx(xj)},${sy(yj)} ${sx(xr)},${sy(yr)}`,
      fill: "none", stroke: "var(--wv-tension)", "stroke-width": 3.5, "stroke-linejoin": "round",
      opacity: 0.9 }));
    // neighbors
    for (const [xx, yy] of [[xl, yl], [xr, yr]]) {
      scene.appendChild(el("circle", { cx: sx(xx), cy: sy(yy), r: r + 1.5,
        fill: "none", stroke: "var(--wv-tension)", "stroke-width": 1.5 }));
    }
    // net transverse force ~ the second difference
    const d2 = (yr - 2 * yj + yl);
    const scale = 2.2;
    const tip = yj + Math.max(-1.6, Math.min(1.6, d2 * scale));
    if (Math.abs(tip - yj) > 0.04) {
      scene.appendChild(el("line", { x1: sx(xj), y1: sy(yj), x2: sx(xj), y2: sy(tip),
        stroke: "var(--wv-force)", "stroke-width": 3, "stroke-linecap": "round" }));
      const up = tip > yj ? 1 : -1;
      scene.appendChild(el("path", {
        d: `M ${sx(xj)} ${sy(tip)} L ${sx(xj) - 5} ${sy(tip - 0.14 * up)} L ${sx(xj) + 5} ${sy(tip - 0.14 * up)} Z`,
        fill: "var(--wv-force)" }));
    }
  }

  readout();
}

function readout() {
  const { N, dx, fidelity } = model;
  document.getElementById("r-t").textContent = state.t.toFixed(2) + " s";
  document.getElementById("r-n").textContent = N;
  document.getElementById("r-dx").textContent = dx.toFixed(3);
  document.getElementById("r-c").textContent = state.c.toFixed(2);
  document.getElementById("r-fid").textContent = (fidelity * 100).toFixed(1) + " %";
}

// ---- derivation stepper ------------------------------------------------
const STEPS = 4;
function showStep() {
  document.querySelectorAll(".step").forEach(d => {
    d.hidden = +d.dataset.step !== state.step;
  });
  document.getElementById("stepno").textContent = state.step + " / " + STEPS;
  document.getElementById("prev").disabled = state.step === 1;
  document.getElementById("next").disabled = state.step === STEPS;
  render();
}
document.getElementById("prev").addEventListener("click", () => {
  if (state.step > 1) { state.step--; showStep(); }
});
document.getElementById("next").addEventListener("click", () => {
  if (state.step < STEPS) { state.step++; showStep(); }
});

// ---- controls ----------------------------------------------------------
const restart = () => { state.t = 0; rebuild(); render(); };
Lab.bind("nn", state, "N", v => String(Math.round(v)), () => { state.N = Math.round(state.N); restart(); });
Lab.bind("cc", state, "c", v => v.toFixed(2), restart);
Lab.bind("pp", state, "pos", v => (v * 100).toFixed(0) + " %", restart);

const play = document.getElementById("play");
play.addEventListener("click", () => {
  state.playing = !state.playing;
  play.textContent = state.playing ? "Pause" : "Play";
});
document.getElementById("replay").addEventListener("click", () => {
  state.t = 0;
  if (!state.playing) render();
});
document.querySelectorAll(".speed button").forEach(b => {
  b.addEventListener("click", () => {
    state.rate = +b.dataset.rate;
    document.querySelectorAll(".speed button")
      .forEach(o => o.setAttribute("aria-pressed", String(o === b)));
  });
});

// drag on the stage to re-pluck there
svg.addEventListener("pointerdown", ev => {
  const r = svg.getBoundingClientRect();
  const x = wx((ev.clientX - r.left) / r.width * VB.w);
  if (x > 0.3 && x < L - 0.3) {
    state.pos = x / L;
    const pp = document.getElementById("pp");
    pp.value = state.pos;
    document.getElementById("pplab").textContent = (state.pos * 100).toFixed(0) + " %";
    restart();
  }
});

// ---- clock -------------------------------------------------------------
let last = null;
function tick(ts) {
  if (last === null) last = ts;
  const dt = Math.min(60, ts - last) / 1000;
  last = ts;
  if (state.playing) {
    state.t += dt * state.rate;
    render();
  }
  requestAnimationFrame(tick);
}

rebuild();
showStep();
requestAnimationFrame(tick);

File: checks/wave.py
"""Wave chain: exact eigenmode evolution, asserted against an
independent Verlet integration, plus the dispersion facts the page states."""
import math

L, c = 10.0, 1.5

def modes(N):
    dx = L / (N + 1)
    return dx, [(2 * c / dx) * math.sin(k * math.pi / (2 * (N + 1))) for k in range(1, N + 1)]

def pluck(N, pos, amp):
    out = []
    for i in range(1, N + 1):
        x = i / (N + 1)
        out.append(amp * (x / pos if x <= pos else (1 - x) / (1 - pos)))
    return out

def dst(y, N):
    return [(2 / (N + 1)) * sum(y[i - 1] * math.sin(k * math.pi * i / (N + 1))
            for i in range(1, N + 1)) for k in range(1, N + 1)]

def evolve(a, N, t, w):
    return [sum(a[k - 1] * math.cos(w[k - 1] * t) * math.sin(k * math.pi * i / (N + 1))
            for k in range(1, N + 1)) for i in range(1, N + 1)]

def run():
    N = 24
    dx, w = modes(N)
    y0 = pluck(N, 0.3, 1.3)
    a = dst(y0, N)

    # DST round-trips exactly
    y_rt = evolve(a, N, 0.0, w)
    assert max(abs(u - v) for u, v in zip(y0, y_rt)) < 1e-12

    # mode evolution == independent Verlet integration
    dt, T_test = 1e-4, 2.0
    y, v = y0[:], [0.0] * N
    def acc(y):
        return [c * c * ((y[i - 1] if i else 0.0) - 2 * y[i] +
                         (y[i + 1] if i < N - 1 else 0.0)) / (dx * dx) for i in range(N)]
    A = acc(y)
    for _ in range(int(T_test / dt)):
        y = [y[i] + v[i] * dt + 0.5 * A[i] * dt * dt for i in range(N)]
        A2 = acc(y)
        v = [v[i] + 0.5 * (A[i] + A2[i]) * dt for i in range(N)]
        A = A2
    err = max(abs(u - q) for u, q in zip(evolve(a, N, T_test, w), y))
    assert err < 1e-6, err

    # energy in mode form is constant (it is, identically -- spot-check)
    def E(t):
        tot = 0.0
        for k in range(1, N + 1):
            q = a[k - 1] * math.cos(w[k - 1] * t)
            qd = -a[k - 1] * w[k - 1] * math.sin(w[k - 1] * t)
            tot += qd * qd + (w[k - 1] * q) ** 2
        return tot
    assert abs(E(0) - E(3.7)) < 1e-9 * E(0)

    # the honest wrinkle: the lattice cutoff pins at 2/pi of c forever --
    # it descends toward 2/pi from above and never gets below it
    ratios = []
    for n in (24, 80, 400):
        dxn, wn = modes(n)
        ratios.append(wn[-1] / (n * math.pi / L) / c)
    lim = 2 / math.pi
    assert ratios[0] > ratios[1] > ratios[2] > lim, ratios
    assert ratios[2] - lim < 2e-3, ratios[2]

    # while the pluck's energy-weighted speed climbs with N
    def fid(n):
        dxn, wn = modes(n)
        an = dst(pluck(n, 0.3, 1.3), n)
        num = den = 0.0
        for k in range(1, n + 1):
            th = k * math.pi / (2 * (n + 1))
            Ek = (an[k - 1] * wn[k - 1]) ** 2
            num += Ek * math.sin(th) / th
            den += Ek
        return num / den
    f5, f24, f80 = fid(5), fid(24), fid(80)
    assert f5 < f24 < f80 < 1.0, (f5, f24, f80)
    return "wave: DST exact, modes==Verlet to %.1e, E const, cutoff->2/pi, fidelity %.3f<%.3f<%.3f" % (err, f5, f24, f80)
